### PR TITLE
[pilot] - reject previous 'waiting for review' build

### LINKED ISSUE #12551: [Testflight] UnexpectedResponse "You can’t submit a build for testing if another build is already in review. Wait until your other build has been reviewed, and resubmit this build."
### New Issue Checklist

- [X] Updated fastlane to the latest version
- [ ] I read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [ ] I read [docs.fastlane.tools](https://docs.fastlane.tools)
- [X] I searched for [existing GitHub issues](https://github.com/fastlane/fastlane/issues)

### Issue Description

If there is the same VersionCode (e.g. 1.2.3) already pushed to testflight and still in testing, fastlane fails with an UnexpectedResponse Error, even tho the build number is different.

Please catch this and make it an option to ignore!

##### Complete output when running fastlane, including the stack trace and command used
<details>
  <pre>
[12:31:55]: Successfully finished processing the build 0.7.7 - 100881
[12:31:59]: Successfully set the changelog for build
[12:31:59]: Distributing new build to testers: 0.7.7 - 100881
+------------------------------------+---------------------------------------------------------------------------------------------------------+
|                                                                 Lane Context                                                                 |
+------------------------------------+---------------------------------------------------------------------------------------------------------+
| DEFAULT_PLATFORM                   | ios                                                                                                     |
| PRODUCE_APPLE_ID                   | 1341311162                                                                                              |
| PLATFORM_NAME                      | ios                                                                                                     |
| LANE_NAME                          | ios alpha                                                                                               |
| BUILD_NUMBER                       | 100881                                                                                                  |
| ORIGINAL_DEFAULT_KEYCHAIN          | "/Users/travis/Library/Keychains/login.keychain-db"                                                     |
| SIGH_PROFILE_TYPE                  | app-store                                                                                               |
| MATCH_PROVISIONING_PROFILE_MAPPING | {"de.democracy-deutschland.clientapp.alpha"=>"match AppStore de.democracy-deutschland.clientapp.alpha"} |
| IPA_OUTPUT_PATH                    | /Users/travis/build/demokratie-live/democracy-development/client/ios/democracyclient.ipa                |
| DSYM_OUTPUT_PATH                   | /Users/travis/build/demokratie-live/democracy-development/client/ios/democracyclient.app.dSYM.zip       |
| XCODEBUILD_ARCHIVE                 | /Users/travis/Library/Developer/Xcode/Archives/2018-05-18/democracyclient 2018-05-18 12.12.22.xcarchive |
+------------------------------------+---------------------------------------------------------------------------------------------------------+
[12:32:00]: {"code"=>"You can’t submit a build for testing if another build is already in review. Wait until your other build has been reviewed, and resubmit this build.", "description"=>"Failed to submit build for review."}
+------+------------------------+-------------+
|              fastlane summary               |
+------+------------------------+-------------+
| Step | Action                 | Time (in s) |
+------+------------------------+-------------+
| 1    | default_platform       | 0           |
| 2    | produce                | 5           |
| 3    | increment_build_number | 3           |
| 4    | automatic_code_signing | 0           |
| 5    | setup_travis           | 0           |
| 6    | sync_code_signing      | 2           |
| 7    | build_app              | 668         |
| 💥   | upload_to_testflight   | 526         |
+------+------------------------+-------------+
[12:32:00]: fastlane finished with errors
Looking for related GitHub issues on fastlane/fastlane...
Found no similar issues. To create a new issue, please visit:
https://github.com/fastlane/fastlane/issues/new
Run `fastlane env` to append the fastlane environment to your issue
bundler: failed to load command: fastlane (/Users/travis/.rvm/gems/ruby-2.4.2/bin/fastlane)
Spaceship::UnexpectedResponse: [!] {"code"=>"You can’t submit a build for testing if another build is already in review. Wait until your other build has been reviewed, and resubmit this build.", "description"=>"Failed to submit build for review."}
  /Users/travis/.rvm/gems/ruby-2.4.2/gems/fastlane-2.95.0/spaceship/lib/spaceship/test_flight/client.rb:299:in `handle_response'
  /Users/travis/.rvm/gems/ruby-2.4.2/gems/fastlane-2.95.0/spaceship/lib/spaceship/test_flight/client.rb:72:in `post_for_testflight_review'
  /Users/travis/.rvm/gems/ruby-2.4.2/gems/fastlane-2.95.0/spaceship/lib/spaceship/test_flight/build.rb:212:in `submit_for_testflight_review!'
  /Users/travis/.rvm/gems/ruby-2.4.2/gems/fastlane-2.95.0/pilot/lib/pilot/build_manager.rb:173:in `distribute_build'
  /Users/travis/.rvm/gems/ruby-2.4.2/gems/fastlane-2.95.0/pilot/lib/pilot/build_manager.rb:99:in `distribute'
  /Users/travis/.rvm/gems/ruby-2.4.2/gems/fastlane-2.95.0/pilot/lib/pilot/build_manager.rb:61:in `upload'
  /Users/travis/.rvm/gems/ruby-2.4.2/gems/fastlane-2.95.0/fastlane/lib/fastlane/actions/upload_to_testflight.rb:16:in `run'
  /Users/travis/.rvm/gems/ruby-2.4.2/gems/fastlane-2.95.0/fastlane/lib/fastlane/runner.rb:257:in `block (2 levels) in execute_action'
  /Users/travis/.rvm/gems/ruby-2.4.2/gems/fastlane-2.95.0/fastlane/lib/fastlane/actions/actions_helper.rb:50:in `execute_action'
  /Users/travis/.rvm/gems/ruby-2.4.2/gems/fastlane-2.95.0/fastlane/lib/fastlane/runner.rb:236:in `block in execute_action'
  /Users/travis/.rvm/gems/ruby-2.4.2/gems/fastlane-2.95.0/fastlane/lib/fastlane/runner.rb:231:in `chdir'
  /Users/travis/.rvm/gems/ruby-2.4.2/gems/fastlane-2.95.0/fastlane/lib/fastlane/runner.rb:231:in `execute_action'
  /Users/travis/.rvm/gems/ruby-2.4.2/gems/fastlane-2.95.0/fastlane/lib/fastlane/runner.rb:157:in `trigger_action_by_name'
  /Users/travis/.rvm/gems/ruby-2.4.2/gems/fastlane-2.95.0/fastlane/lib/fastlane/fast_file.rb:149:in `method_missing'
  Fastfile:65:in `block (2 levels) in parsing_binding'
  /Users/travis/.rvm/gems/ruby-2.4.2/gems/fastlane-2.95.0/fastlane/lib/fastlane/lane.rb:33:in `call'
  /Users/travis/.rvm/gems/ruby-2.4.2/gems/fastlane-2.95.0/fastlane/lib/fastlane/runner.rb:49:in `block in execute'
  /Users/travis/.rvm/gems/ruby-2.4.2/gems/fastlane-2.95.0/fastlane/lib/fastlane/runner.rb:45:in `chdir'
  /Users/travis/.rvm/gems/ruby-2.4.2/gems/fastlane-2.95.0/fastlane/lib/fastlane/runner.rb:45:in `execute'
  /Users/travis/.rvm/gems/ruby-2.4.2/gems/fastlane-2.95.0/fastlane/lib/fastlane/lane_manager.rb:59:in `cruise_lane'
  /Users/travis/.rvm/gems/ruby-2.4.2/gems/fastlane-2.95.0/fastlane/lib/fastlane/command_line_handler.rb:36:in `handle'
  /Users/travis/.rvm/gems/ruby-2.4.2/gems/fastlane-2.95.0/fastlane/lib/fastlane/commands_generator.rb:107:in `block (2 levels) in run'
  /Users/travis/.rvm/gems/ruby-2.4.2/gems/commander-fastlane-4.4.6/lib/commander/command.rb:178:in `call'
  /Users/travis/.rvm/gems/ruby-2.4.2/gems/commander-fastlane-4.4.6/lib/commander/command.rb:153:in `run'
  /Users/travis/.rvm/gems/ruby-2.4.2/gems/commander-fastlane-4.4.6/lib/commander/runner.rb:476:in `run_active_command'
  /Users/travis/.rvm/gems/ruby-2.4.2/gems/fastlane-2.95.0/fastlane_core/lib/fastlane_core/ui/fastlane_runner.rb:75:in `run!'
  /Users/travis/.rvm/gems/ruby-2.4.2/gems/commander-fastlane-4.4.6/lib/commander/delegates.rb:15:in `run!'
  /Users/travis/.rvm/gems/ruby-2.4.2/gems/fastlane-2.95.0/fastlane/lib/fastlane/commands_generator.rb:332:in `run'
  /Users/travis/.rvm/gems/ruby-2.4.2/gems/fastlane-2.95.0/fastlane/lib/fastlane/commands_generator.rb:41:in `start'
  /Users/travis/.rvm/gems/ruby-2.4.2/gems/fastlane-2.95.0/fastlane/lib/fastlane/cli_tools_distributor.rb:108:in `take_off'
  /Users/travis/.rvm/gems/ruby-2.4.2/gems/fastlane-2.95.0/bin/fastlane:20:in `<top (required)>'
  /Users/travis/.rvm/gems/ruby-2.4.2/bin/fastlane:23:in `load'
  /Users/travis/.rvm/gems/ruby-2.4.2/bin/fastlane:23:in `<top (required)>'
The command "DELIVER_ITMSTRANSPORTER_ADDITIONAL_UPLOAD_PARAMETERS="-t DAV" bundle exec fastlane ios alpha" exited with 1.
  </pre>
</details>

### Environment

https://travis-ci.org/demokratie-live/democracy-development/jobs/380659794

### HowToTest

- Push App to Testflight with a specific Version Code (e.g. 1.2.3)
- Push the same Version with a new Build Number immediately(<= 6h) after the first Push.
- ...
- Profit

### PR DESCRIPTION
**Pull request by RishabhTayal**
fixes: #12551

### CODE REVIEW

File: pilot/lib/pilot/options.rb [hunk 1]
```diff
@@ -189,6 +189,12 @@ def self.available_options
                                      env_name: "PILOT_WAIT_FOR_UPLOADED_BUILD",
                                      description: "Use version info from uploaded ipa file to determine what build to use for distribution. If set to false, latest processing or any latest build will be used",
                                      is_string: false,
+                                     default_value: false),
+        FastlaneCore::ConfigItem.new(key: :reject_previously_submitted_build,
```
Comment (RishabhTayal) [inline pilot/lib/pilot/options.rb]: Not a perfect name. Looking for suggestions to improve this option name

**Comment by RishabhTayal** [comment]
@joshdholtz This is ready for review.

**Comment by ulfgebhardt** [comment]
Please merge. Confirming fix


image


Confirming that this works.

Great job @RishabhTayal & @ReggaePanda 

<3

Rel: https://github.com/fastlane/fastlane/issues/12551

**Review by joshdholtz — CHANGES_REQUESTED** [review]
@RishabhTayal Hey, this is a great addition! Is it possible though to use your `all_waiting_for_review` to reject a build before `distribute_build` is called? Checking for strings in the error response is a bit unreliable. It would be better to query for  the "waiting for review" state first, reject, and then distribution. This would also get rid of some potentially dangerous recursion 😉

**Comment by RishabhTayal** [comment]
@joshdholtz Great suggestion! I have moved the logic to reject build before distribute method. This is ready for review again.

Thanks.

File: pilot/lib/pilot/build_manager.rb [hunk 2]
```diff
@@ -96,6 +96,15 @@ def distribute(options, build: nil)
       build.auto_notify_enabled = config[:notify_external_testers]
 
       return if config[:skip_submission]
+      if options[:reject_previously_submitted_build]
+        waiting_for_review_build = Spaceship::TestFlight::Build.all_waiting_for_review(app_id: build.app_id, platform: :ios).first
```
Comment (joshdholtz) [inline pilot/lib/pilot/build_manager.rb]: Do we want to throw an error here if there is more than one? 🤔 Calling `.first` seems save enough but would be confusing **if** there was two and then it failed for this reason 😬

File: pilot/lib/pilot/options.rb (hunk 1 above)
Comment (joshdholtz) [reply pilot/lib/pilot/options.rb]: Can we name this `:reject_build_waiting_for_review`? This seems to be a bit more specific to what its doing 😊

**Review by joshdholtz — CHANGES_REQUESTED** [review]
Two small comments 😊

File: pilot/lib/pilot/build_manager.rb (hunk 2 above)
Comment (RishabhTayal) [reply pilot/lib/pilot/build_manager.rb]: @joshdholtz I don't think there could be more than one 'waiting for review'. Apple doesn't allow more than one build to be submitted for review. So this array would always have max 1 value.

**Comment by RishabhTayal** [comment]
@joshdholtz Made the option name change. ready for review again.

File: spaceship/lib/spaceship/test_flight/build.rb [hunk 3]
```diff
@@ -119,6 +119,11 @@ def self.all_processing_builds(app_id: nil, platform: nil, retry_count: 0)
         all(app_id: app_id, platform: platform, retry_count: retry_count).find_all(&:processing?)
       end
 
+      # Just the builds, as a flat array, that are waiting for beta review
+      def self.all_waiting_for_review(app_id: nil, platform: nil, retry_count: 0)
```
Comment (heartofagoof) [inline spaceship/lib/spaceship/test_flight/build.rb]: Is it possible for multiple builds to be waiting for review?

Comment (RishabhTayal) [reply spaceship/lib/spaceship/test_flight/build.rb]: https://github.com/fastlane/fastlane/pull/12565#discussion_r192190395

Comment (heartofagoof) [reply spaceship/lib/spaceship/test_flight/build.rb]: Nice! That is my understanding as well. I was just thinking that `all` in `all_waiting_for_review` can throw someone off. may be indicate that there can only be one by the method name too?

**Comment by ReggaePanda** [comment]
@joshdholtz can u merge it? 😃

**Comment by joshdholtz** [comment]
@ReggaePanda Sorry, was out of town all last week! Looking and testing this now

File: pilot/lib/pilot/build_manager.rb [hunk 4]
```diff
@@ -96,6 +96,15 @@ def distribute(options, build: nil)
       build.auto_notify_enabled = config[:notify_external_testers]
 
       return if config[:skip_submission]
+      if options[:reject_build_waiting_for_review]
+        waiting_for_review_build = Spaceship::TestFlight::Build.all_waiting_for_review(app_id: build.app_id, platform: :ios).first
```
Comment (joshdholtz) [inline pilot/lib/pilot/build_manager.rb]: Should this be `platform: fetch_app_platform`?

**Review by joshdholtz — CHANGES_REQUESTED** [review]
Tested this out and it works 💯 Just one quick questions regarding a hard coded `:ios`

File: pilot/lib/pilot/build_manager.rb (hunk 4 above)
Comment (RishabhTayal) [reply pilot/lib/pilot/build_manager.rb]: @joshdholtz Good catch. I have made the required change. 👍

**Comment by RishabhTayal** [comment]
@joshdholtz This is ready for review again.

**Review by joshdholtz — APPROVED** [review]
This feels solid to me 👍

**Pull request merged**